File: JavaCivitas/src/civitas/TituloPropiedad.java
package civitas;

public class TituloPropiedad {
    
    private float alquilerBase;
    private static float factorInteresesHipoteca = (float) 1.1;
    private float factorRevalorizacion;
    private float hipotecaBase;
    private boolean hipotecado;
    private String nombre;
    private int numCasas;
    private int numHoteles;
    private float precioCompra;
    private float precioEdificar;
    
    
    //añadido propietario
    private Jugador propietario;
    
    void actualiarPropietarioPorConversion(Jugador jugador){
        propietario = jugador;
    }
    
    boolean cancelarHipoteca(Jugador jugador){
        if(hipotecado && esEsteElPropietario(jugador)){
            if(jugador.paga(getImporteCancelarHipoteca())){
                hipotecado = false;
                return true;
            }else{
                return false;
            }
        }else{
            return false;
        }
    }

    public float getPrecioEdificar() {
        return precioEdificar;
    }

    public String getNombre() {
        return nombre;
    }

    public TituloPropiedad(String nom, float ab, float fr, float hb, float pc, float pe){

        nombre = nom;
        alquilerBase = ab;
        factorRevalorizacion = fr;
        hipotecaBase = hb;
        precioCompra = pc;
        precioEdificar = pe;
        hipotecado=false;
        numCasas=0;
        numHoteles=0;

    }
    @Override
    public String toString(){

        return "\n----------[PROPIEDAD]----------\n[Nombre] -> " + nombre
        + "\n[Alquiler base] -> " + alquilerBase + "\n[Factor InteresesHipoteca] -> " + factorInteresesHipoteca
        + "\n[Factor Revalorizacion] -> " + factorRevalorizacion + "\n[Hipoteca base] -> " + hipotecaBase
        + "\n[Hipotecado] -> " + hipotecado + "\n[Numero casas] -> " + numCasas + "\n[Numero Hoteles] -> " + numHoteles
        + "\n[Precio compra] -> " + precioCompra + "\n[Precio edificacion] ->  " + precioEdificar + "\n-------------------------------";

    }

    public float getPrecioAlquiler(){

        if(hipotecado || propietarioEncarcelado()){// añadido propietario es encarcelado

            return 0;

        }else{

            return (float) (alquilerBase*(1+(numCasas*0.5) + (numHoteles*2.5)));

        }
    }
    
    private float getImporteHipoteca(){   
        return (float) (hipotecaBase*(1+(numCasas*0.5)+(numHoteles*2.5)));
    }

    float getImporteCancelarHipoteca(){

        // (getImporteHipoteca() *factorInteresesHipoteca);
        
        return (getImporteHipoteca() * factorInteresesHipoteca);
        

    }

    void tramitarAlquiler(Jugador jugador){

        if((getPropietario() != null) && (propietario != jugador)){

            //float alquiler = jugador.pagaAlquiler();
            //getPropietario().recibe(alquiler);
            
            jugador.pagaAlquiler(alquilerBase);
            propietario.recibe(getPrecioAlquiler());
            

        }

    }

    private boolean propietarioEncarcelado(){

        if(tienePropietario()){

            if(propietario.isEncarcelado()){
                return true;
            }else{
                return false;
            }

        }else{
            return false;
        }

    }

    int cantidadCasasHoteles(){

        return numHoteles+numCasas;

    }

    private float getPrecioVenta(){

        return precioCompra + (precioEdificar*numCasas) + (precioEdificar*numHoteles*factorRevalorizacion);

    }
    int getNumCasas(){
        return numCasas;
    }
    int getNumHoteles(){
        return numHoteles;
    }
    
    float getPrecioCompra(){
        return precioCompra;
    }
    
    boolean comprar(Jugador jugador){

        if (propietario == null){
            propietario = jugador;
            propietario.paga(precioCompra);
            return true;

        }

        return false;
    }
    boolean construirCasa(Jugador jugador){
        if (esEsteElPropietario(jugador)){
            if(jugador.paga(getPrecioEdificar())){
                numCasas++;
                return true; 
            }else{
                return false;
            } 
        }else{
            return false;
        }
            
    }
    boolean construirHotel(Jugador jugador){
        if (esEsteElPropietario(jugador)){
            if(jugador.paga(getPrecioEdificar())){
                numHoteles++;
                numCasas = 0;
                return true; 
            }else{
                return false;
            }   
        }else{
            return false;
        }             
    }

    boolean derruirCasas(int n, Jugador jugador){

        if(jugador == getPropietario() && numCasas >= n){

            numCasas -= n;
            return true;

        }else{
            return false;
        }

    }
    boolean esEsteElPropietario(Jugador jugador){
        if (jugador == propietario){
            return true;
        }else{
            return false;
        }
    }

    boolean vender(Jugador jugador){

        if(getPropietario() == jugador && !hipotecado){
            if(propietario.recibe(getPrecioVenta())){
                propietario=jugador;
                numCasas = 0;
                numHoteles = 0;
                return true;
            }else{
                return false;
            }
        }else{
            return false;
        }
    }
    
    
    
    
    
    Jugador getPropietario(){
        return propietario;
    }
    
    public boolean getHipotecado(){
        return this.hipotecado;
    }
    boolean tienePropietario(){
        if(hipotecado){
            return true;
        }else{
            return false;
        }
    }
    
    boolean hipotecar(Jugador jugador){
        if (!hipotecado && esEsteElPropietario(jugador)){
            if(jugador.recibe(getImporteHipoteca())){
            hipotecado= true;
            return true;
            }else{
                return false;
            }
        }else{
            return false;
        }       
    }
    
    



}

File: RubyCivitas/pkg/RubyApplication1-0.0.1/lib/titulo_propiedad.rb
require_relative "jugador"
module Civitas
  class TituloPropiedad

    @@factor_intereses_hipoteca = 1.1

    def initialize(nom, ab, fr, hb, pc, pe)
      @alquiler_base = ab
      @factor_revalorizacion = fr
      @hipoteca_base = hb
      @hipotecado = false
      @nombre = nom
      @num_casas = 0
      @num_hoteles = 0
      @precio_compra = pc
      @precio_edificar = pe
      @propietario = nil
    end

    def to_string

      "----------[PROPIEDAD]----------\n[Nombre] -> " + @nombre.to_s
      + "\n[Alquiler base] -> " + @alquiler_base.to_s + "\n[Factor InteresesHipoteca] -> " + @@factor_intereses_hipoteca.to_s
      + "\n[Factor Revalorizacion] -> " + @factor_revalorizacion.to_s + "\n[Hipoteca base] -> " + @hipoteca_base.to_s
      + "\n[Hipotecado] -> " + @hipotecado.to_s + "\n[Numero casas] -> " + @num_casas.to_s + "\n[Numero Hoteles] -> " + @num_hoteles.to_s
      + "\n[Precio compra] -> " + @precio_compra.to_s + "\n[Precio edificacion] ->  " + @precio_edificar.to_s + "\n-------------------------------"

    end

    def get_precio_alquiler

      if @hipotecado
        0
      else

        (@alquiler_base * (1 + (@num_casas * 0.5) + (@num_hoteles * 2.5)))

      end
    end

    def tiene_propietario

      !@propietario.nil?

    end

    def get_importe_hipoteca

      @hipoteca_base*((@num_casas*0.5) + 1 + (2.5*@num_hoteles))

    end

    def es_este_el_propietario(jugador)

      @propietario == jugador

    end

    def get_importe_cancelar_hipoteca

      get_importe_hipoteca * @@factor_intereses_hipoteca
    end

    def cancelar_hipoteca(jugador)

      false

      if @hipotecado && es_este_el_propietario(jugador)

        jugador.paga(get_importe_cancelar_hipoteca)
        @hipotecado = false
        true

      end
    end

    def tramitar_alquiler(jugador)

      if es_este_el_propietario(jugador)

        jugador.paga_alquiler(get_precio_alquiler)
        @propietario.recibe(get_precio_alquiler)

      end
    end

    def propietario_encarcelado
      tiene_propietario && @propietario.is_encarcelado
    end

    def cantidad_casas_hoteles

      @num_hoteles + @num_casas

    end

    def get_precio_venta

      @precio_compra + (@precio_edificar * @num_casas) + (@precio_edificar * @num_hoteles * @factor_revalorizacion)

    end

    def derruir_casas(n, jugador)

      if es_este_el_propietario(jugador) && @num_casas >= n
        @num_casas -= n
        return true
      end
      false
    end

    def vender(jugador)

      if es_este_el_propietario(jugador) && !@hipotecado
        jugador.recibe(get_precio_venta)
        @propietario = nil
        @num_casas = 0
        @num_hoteles = 0
        return true
      end

      false
    end
    
    def comprar(jugador)
      
      final = false
      if @propietario == nil
        @propietario = jugador
        @propietario.paga(@precio_compra)
        final = true
      end
      
      return final
      
    end

    def construir_hotel(jugador)

      construido = false

      if es_este_el_propietario(jugador)

        jugador.paga(@precio_edificar)
        @num_hoteles += 1
        construido = true
      end

      construido
    end

    def construir_casa(jugador)

      decision = false

      if es_este_el_propietario(jugador)

        jugador.paga(@precio_edificar)
        @num_casas += 1
        decision = true

      end
    end

    def hipotecar(jugador)

      decision = false

      if !@hipotecado && es_este_el_propietario(jugador)

        jugador.recibe(get_importe_hipoteca)
        @hipotecado = true
        decision = true

      end
    end

    attr_reader :hipotecado, :hipoteca_base, :nombre, :num_casas, :num_hoteles,
      :alquiler_base, :precio_compra, :precio_edificar

  end
end
